src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -267,6 +279,65 @@ private IEnumerable<string> GetAttributeNames(Block tagBlock)
             return attributeNames;
         }
 
+        private void ValidateParentTagHelperAllowsContent(SyntaxTreeNode child, ErrorSink errorSink)
+        {
+            ValidateAllowedChildren(
+                RazorResources.TagHelperParseTreeRewriter_CannotHaveNonTagContent,
+                child,
+                errorSink);
+        }
+
+        private void ValidateParentTagHelperAllowsPlainTag(Block tagBlock, ErrorSink errorSink)
+        {
+            ValidateAllowedChildren(
+                RazorResources.TagHelperParseTreeRewriter_InvalidNestedTag,
+                tagBlock,
+                errorSink);
+        }
+
+        private void ValidateParentTagHelperAllowsTagHelper(string tagName, Block tagBlock, ErrorSink errorSink)
+        {
+            var currentTracker = _trackerStack.Count > 0 ? _trackerStack.Peek() : null;
+            var currentlyAllowedChildren = currentTracker?.AllowedChildren;
+
+            if (currentlyAllowedChildren != null &&
+                !currentlyAllowedChildren.Contains(tagName, StringComparer.OrdinalIgnoreCase))
+            {
+                OnAllowedChildrenError(
+                    currentTracker,
+                    RazorResources.TagHelperParseTreeRewriter_InvalidNestedTag,
+                    tagBlock,
+                    errorSink);
+            }
+        }
+
+        private void ValidateAllowedChildren(string errorMessageFormat, SyntaxTreeNode node, ErrorSink errorSink)
+        {
+            var currentTracker = _trackerStack.Count > 0 ? _trackerStack.Peek() : null;
+            var currentlyAllowedChildren = currentTracker?.AllowedChildren;
+
+            if (currentlyAllowedChildren != null)
```

Comment:
This gets called a few times per encountered tag name and numerous times per parent tag. Suggest maintaining a `_currentlyAllowedChildren` field when the `_trackerStack` is changed would be more straightforward.